autogpt/memory/vector/providers/base.py

```diff
@@ -0,0 +1,68 @@
+import functools
+from typing import MutableSet, Sequence
+
+import numpy as np
+
+from autogpt.logs import logger
+from autogpt.singleton import AbstractSingleton
+
+from .. import MemoryItem, MemoryItemRelevance
+from ..utils import Embedding, get_embedding
+
+
+class VectorMemoryProvider(MutableSet[MemoryItem], AbstractSingleton):
```

Comment:
This is not an interface.  This is a quite specific implementation primarily thinking about using a local json backend.  It is also delegating tasks to MemoryItem/MemoryItemRelevance that it should be doing itself (making embeddings and doing summarization, though I would probably argue that doing summarization is outside the scope of this abstraction).

Reply:
It is not an interface: it is an abstract with a couple of generic method implementations (`get`, `get_relevant`, `get_stats`) which utilize the storage interface that every backend should expose (`MutableSet`). Memory store implementations may but don't have to override these generic methods.

I consciously chose `MutableSet` as the basis of the memory storage interface. Feel free to argue against that choice if you see problems with it.

Reply:
Okay, good for now, will talk more about it tomorrow

Reply:
Todo: talk about this